app/scripts/controllers/authentication/authentication-controller.ts

```diff
@@ -0,0 +1,247 @@
+  AuthenticationControllerMessenger
+> {
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: AuthenticationControllerMessenger;
+    state?: AuthenticationControllerState;
+  }) {
+    super({
+      messenger,
+      metadata,
+      name: controllerName,
+      state: { ...defaultState, ...state },
+    });
+  }
+
+  public async performSignIn(): Promise<string> {
+    const { accessToken } = await this.#performAuthenticationFlow();
+    return accessToken;
+  }
+
+  public performSignOut(): void {
+    this.#assertLoggedIn();
+
+    this.update((state) => {
+      state.isSignedIn = false;
+      state.sessionData = undefined;
+    });
+  }
+
+  public async getBearerToken(): Promise<string> {
+    this.#assertLoggedIn();
+
+    if (this.#hasValidAuthTokens(this.state.sessionData)) {
+      return this.state.sessionData.accessToken;
+    }
+
+    const { accessToken } = await this.#performAuthenticationFlow();
+    return accessToken;
+  }
+
+  /**
+   * NOTE this will be changed to use profileId in future task.
+   *
+   * @returns the identifier id
+   */
+  public async getIdentifierId(): Promise<string> {
+    this.#assertLoggedIn();
+
+    if (this.#hasValidAuthTokens(this.state.sessionData)) {
+      return this.state.sessionData?.profile?.identifierId;
+    }
+
+    const { identifierId } = await this.#performAuthenticationFlow();
+    return identifierId;
+  }
+
+  #assertLoggedIn() {
+    if (!this.state.isSignedIn) {
+      throw new Error(
+        `${controllerName}: Unable to call method, user is not authenticated`,
+      );
+    }
+  }
+
+  async #performAuthenticationFlow() {
+    try {
+      // 1. Nonce
+      const publicKey = await this.#snapGetPublicKey();
+      const nonce = await getNonce(publicKey);
+      if (!nonce) {
+        throw new Error(`Unable to get nonce`);
+      }
+
+      // 2. Login
+      const rawMessage = createLoginRawMessage(nonce, publicKey);
+      const signature = await this.#snapSignMessage(rawMessage);
+      const loginResponse = await login(rawMessage, signature);
+      if (!loginResponse?.token) {
+        throw new Error(`Unable to login`);
+      }
+
+      // 3. Trade for Access Token
+      const accessToken = await getAccessToken(loginResponse.token);
+      if (!accessToken) {
+        throw new Error(`Unable to get Access Token`);
+      }
+
+      // Update Internal State
+      this.update((state) => {
+        state.isSignedIn = true;
+        const expiresIn = new Date();
+        expiresIn.setTime(expiresIn.getTime() + THIRTY_MIN_MS);
+        state.sessionData = {
+          profile: { identifierId: loginResponse.profile.identifier_id },
+          accessToken,
+          expiresIn: expiresIn.toString(),
+        };
+      });
+
+      return {
+        identifierId: loginResponse.profile.identifier_id,
+        accessToken,
+      };
+    } catch (e) {
+      throw new Error(`${controllerName}: Failed to authenticate - ${e}`);
```

Comment:
I haven't tested it but I think we would see [object Object] here. Perhaps we can JSON.stringify the error or pass the message only?

Reply:
Great point. Tested this.

Since we are using Errors, they will automatically use the `.toString()` when using string concatenation.

But just for safety, I'll add a guard in case if it isn't an `instanceof Error`.

```
const errorMessage = e instanceof Error ? e.toString() : JSON.stringify(e)
```